### ISSUE TITLE
Steps for Running Model on My Own Video

### ISSUE THREAD

**Issue by keertikbacon** [JaywongWang/DenseVideoCaptioning#26]
I would like to use your model to caption a video of my own, and so I just want to clarify the steps I need to take to do so. Please correct me if I'm wrong.

1. Extract C3D features from the video (using the code from [this example](https://github.com/facebook/C3D/tree/master/C3D-v1.0/examples/c3d_feature_extraction))

2. Convert the features into numpy array files

3. Convert those into an h5py file using [this program](https://github.com/JaywongWang/DenseVideoCaptioning/blob/master/dataset/ActivityNet_Captions/features/get_c3d_h5.py)

4. Run test.py, using the pretrained model, replacing test_ids with my own, and setting features to the h5py file I just created

I tried to piece the process together using some of the other issues, but I still had doubts, so I just wanted to clarify it.

**Comment by JaywongWang** [comment]
The first step (feature extraction) should be followed by PCA processing (4096 -> 500). Otherwise the mismatched feature dimension would cause runtime error. Please follow http://activity-net.org/challenges/2016/download.html#c3d.

**Comment by baconsuperguy** [comment]
Thanks for the reply.

For PCA processing down to 500 dimensions, it seems you must first start with at least 500 data points. So I'm not quite sure how to PCA process the C3D features I've extracted from one video. Is there something I'm missing?

**Comment by keertikbacon** [comment]
Does ActivityNet provide the PCA model that they used to process their data? I can't seem to find it. I did find the "PCA_activitynet_v1-3.hdf5" file, is this the PCA model? If it is, how do I use it to process my test data?

**Comment by JaywongWang** [comment]
PCA_activitynet_v1-3.hdf5 provides information needed to perform PCA on new data points (standardization, projection, etc.). As this is beyond the scope of thie repo, please refer to other online guidances.

**Comment by keertikbacon** [comment]
Thanks, I was able to get it working.

**Comment by YuliaNovskaya** [comment]
> Thanks, I was able to get it working.

Could you please explain how did you do it?

**Comment by ebron01** [comment]
@baconsuperguy how did you manage to do PCA on a different video dataset?

**Comment by keertikbacon** [comment]
I used the PCA_activitynet_v1-3.hdf5 file I talked about previously. First, I subtracted the "x_mean" value from my video's C3D features. And then, I took the first 500 components of the "U" value and I did a dot product with the C3D features.

I got help from others in figuring out how to do this, so I unfortunately don't know enough to answer any questions you might have. But I hope I helped!

**Comment by aurooj** [comment]
Hello, I wanted to know one thing. Do you save features for each video clip as a separate numpy or one numpy file with all features. 
For example, if my video clip has 7 mini-clips i.e. 7 c3d features, after I apply PCA, I will get (7, 500) dimensional feature for my full video clip. Can the model take the batch of such features (shape would be (batch_size, #clips, 500)) as an input?

**Comment by aurooj** [comment]
@baconsuperguy , thanks for your reply. I have computed PCA on my dataset the way you told and saved features in an hdf5 file. 
However I am stuck at the last step now. Since it throws the following error when loading pretrained model:
Restoring model from checkpoints/model.ckpt
W tensorflow/core/framework/op_kernel.cc:993] Not found: Tensor name "caption_module/multi_rnn_cell/cell_0/lstm_cell/weights" not found in checkpoint files checkpoints/model.ckpt
	 [[Node: save/RestoreV2_11 = RestoreV2[dtypes=[DT_FLOAT], _device="/job:localhost/replica:0/task:0/cpu:0"](_recv_save/Const_0, save/RestoreV2_11/tensor_names, save/RestoreV2_11/shape_and_slices)]]
W tensorflow/core/framework/op_kernel.cc:993] Not found: Tensor name "caption_module/multi_rnn_cell/cell_0/lstm_cell/weights" not found in checkpoint files checkpoints/model.ckpt....

at line 79 in test.py: 
saver.restore(sess, options['init_from'])

Your help will save my day :(

**Comment by a1391651300** [comment]
> I used the PCA_activitynet_v1-3.hdf5 file I talked about previously. First, I subtracted the "x_mean" value from my video's C3D features. And then, I took the first 500 components of the "U" value and I did a dot product with the C3D features.
> 
> I got help from others in figuring out how to do this, so I unfortunately don't know enough to answer any questions you might have. But I hope I helped!

Could you please show you code for C3D and pca? I am a beginner

**Comment by a1391651300** [comment]
> @baconsuperguy , thanks for your reply. I have computed PCA on my dataset the way you told and saved features in an hdf5 file.
> However I am stuck at the last step now. Since it throws the following error when loading pretrained model:
> Restoring model from checkpoints/model.ckpt
> W tensorflow/core/framework/op_kernel.cc:993] Not found: Tensor name "caption_module/multi_rnn_cell/cell_0/lstm_cell/weights" not found in checkpoint files checkpoints/model.ckpt
> [[Node: save/RestoreV2_11 = RestoreV2[dtypes=[DT_FLOAT], _device="/job:localhost/replica:0/task:0/cpu:0"](_recv_save/Const_0, save/RestoreV2_11/tensor_names, save/RestoreV2_11/shape_and_slices)]]
> W tensorflow/core/framework/op_kernel.cc:993] Not found: Tensor name "caption_module/multi_rnn_cell/cell_0/lstm_cell/weights" not found in checkpoint files checkpoints/model.ckpt....
> 
> at line 79 in test.py:
> saver.restore(sess, options['init_from'])
> 
> Your help will save my day :(

I solve the problem by upgrade the tensorflow to 1.7. Now can you run your own video with the project? Could you show me the method or code,best tanks.

**Comment by aurooj** [comment]
@a1391651300 
Hay, I was able to solve the problem and yes, you are right, it was solely a version issue. 
Here is how I ran PCA on my C3D features. PLease note that I already had saved c3d features as numpy files in my case.

`import os
import numpy as np
import h5py
from tqdm import tqdm

#takes about 20-30 minutes to do PCA on features

PCA_activitynet = h5py.File('dataset/ActivityNet_Captions/PCA_activitynet_v1-3.hdf5', 'r')

x_mean = PCA_activitynet['data/x_mean']
U = PCA_activitynet['data/U'][:500]
U = np.expand_dims(U, axis=2)

#read c3d features for your dataset
root = 'path/to/c3d/feat/saved/as/numpy/files'

save_dir = "path/to/folder/to/save/feats/after/PCA"

#iterate through feature numpy files and save them after PCA

dirs = os.listdir(root)
for dir_ in dirs:
    video_clips = os.listdir(os.path.join(root, dir_))

    for clip in tqdm(video_clips):
        sub_clips = os.listdir(os.path.join(root, dir_, clip))

        save_path = os.path.join(save_dir, dir_, clip)
        if not os.path.isdir(save_path):
            os.makedirs(save_path)

        for sub_clip in sub_clips:
            np_feat = np.load(os.path.join(root, dir_, clip, sub_clip))
            diff_feat = np_feat - x_mean #subtrac PCA mean from each feature

            diff_feat = np.expand_dims(diff_feat, axis=1)
            dot_prod = np.squeeze(np.dot(diff_feat, U))
            if len(dot_prod.shape) == 1:
                dot_prod = np.expand_dims(dot_prod, axis=0)
            #print(dot_prod.shape)
            np.save(os.path.join(save_path, sub_clip), dot_prod)`
